assembly/src/parser/grammar.lalrpop

```diff
@@ -0,0 +1,1379 @@
+        }
+    },
+    <l:@L> "u32shr" <imm:MaybeImm<Shift32>> <r:@R> => {
+        let span = span!(l, r);
+        match imm {
+            Some(imm) if imm == 0 => smallvec![],
+            Some(imm) => smallvec![Op::Inst(Span::new(span, Instruction::U32ShrImm(imm)))],
+            None => smallvec![Op::Inst(Span::new(span, Instruction::U32Shr))],
+        }
+    },
+    <l:@L> "u32rotl" <imm:MaybeImm<Shift32>> <r:@R> => {
+        let span = span!(l, r);
+        match imm {
+            Some(imm) if imm == 0 => smallvec![],
+            Some(imm) => smallvec![Op::Inst(Span::new(span, Instruction::U32RotlImm(imm)))],
+            None => smallvec![Op::Inst(Span::new(span, Instruction::U32Rotl))],
+        }
+    },
+    <l:@L> "u32rotr" <imm:MaybeImm<Shift32>> <r:@R> => {
+        let span = span!(l, r);
+        match imm {
+            Some(imm) if imm == 0 => smallvec![],
+            Some(imm) => smallvec![Op::Inst(Span::new(span, Instruction::U32RotrImm(imm)))],
+            None => smallvec![Op::Inst(Span::new(span, Instruction::U32Rotr))],
+        }
+    },
+}
+
+#[inline]
+InstWithFeltImmediate: Instruction = {
+    "exp" <imm:MaybeImm<Felt>> => imm.map(Instruction::ExpImm).unwrap_or(Instruction::Exp),
+}
+
+#[inline]
+InstWithU32Immediate: Instruction = {
+    "mem_load" <imm:MaybeImm<U32>> => imm.map(Instruction::MemLoadImm).unwrap_or(Instruction::MemLoad),
+    "mem_loadw" <imm:MaybeImm<U32>> => imm.map(Instruction::MemLoadWImm).unwrap_or(Instruction::MemLoadW),
+    "mem_store" <imm:MaybeImm<U32>> => imm.map(Instruction::MemStoreImm).unwrap_or(Instruction::MemStore),
+    "mem_storew" <imm:MaybeImm<U32>> => imm.map(Instruction::MemStoreWImm).unwrap_or(Instruction::MemStoreW)
+}
+
+#[inline]
+InstWithLocalIndex: Instruction = {
+    "locaddr" <i:Imm<U16>> => Instruction::Locaddr(i),
+    "loc_load" <i:Imm<U16>> => Instruction::LocLoad(i),
+    "loc_loadw" <i:Imm<U16>> => Instruction::LocLoadW(i),
+    "loc_store" <i:Imm<U16>> => Instruction::LocStore(i),
+    "loc_storew" <i:Imm<U16>> => Instruction::LocStoreW(i)
+}
+
+#[inline]
+InstWithStackIndex: Instruction = {
+    "adv_push" "." <l:@L> <i:OneBasedStackIndex> <r:@R> => Instruction::AdvPush(Immediate::Value(Span::new(span!(l, r), i))),
+    "dup" <i:("." <ZeroBasedStackIndex>)?> => match i.unwrap_or(0) {
+        0 => Instruction::Dup0,
+        1 => Instruction::Dup1,
+        2 => Instruction::Dup2,
+        3 => Instruction::Dup3,
+        4 => Instruction::Dup4,
+        5 => Instruction::Dup5,
+        6 => Instruction::Dup6,
+        7 => Instruction::Dup7,
+        8 => Instruction::Dup8,
+        9 => Instruction::Dup9,
+        10 => Instruction::Dup10,
+        11 => Instruction::Dup11,
+        12 => Instruction::Dup12,
+        13 => Instruction::Dup13,
+        14 => Instruction::Dup14,
+        15 => Instruction::Dup15,
+        _ => unreachable!(),
+    },
+    "dupw" <i:("." <ZeroBasedWordIndex>)?> => match i.unwrap_or(0) {
```

Comment:
`ZeroBasedWordIndex` seems to be used only here, Myabe replace it with `<U8>` like it is done for the `movdn` and have the parsing error instead of the unreachable? (It makes it easier to check the code)

Reply:
Agreed, I had originally expected it to be used in a few more places, but it didn't actually play out that way, so this reduces the size of the grammar a bit at no real cost in legibility (it may actually be easier, as you suggest).